messages.go

```diff
@@ -0,0 +1,140 @@
+package dht
+
+import (
+	"bytes"
+	"context"
+	"errors"
+	"fmt"
+	"github.com/libp2p/go-libp2p-core/host"
+	"github.com/libp2p/go-libp2p-core/protocol"
+	"github.com/libp2p/go-libp2p-core/routing"
+	record "github.com/libp2p/go-libp2p-record"
+	"github.com/multiformats/go-multihash"
+
+	"github.com/libp2p/go-libp2p-core/peer"
+
+	recpb "github.com/libp2p/go-libp2p-record/pb"
+
+	pb "github.com/libp2p/go-libp2p-kad-dht/pb"
+)
+
+type ProtocolMessenger struct {
+	m         *messageManager
+	validator record.Validator
+}
+
+func NewProtocolMessenger(host host.Host, protocols []protocol.ID, validator record.Validator) *ProtocolMessenger {
+	return &ProtocolMessenger{
+		m: &messageManager{
+			host:      host,
+			strmap:    make(map[peer.ID]*messageSender),
+			protocols: protocols,
+		},
+		validator: validator,
+	}
+}
+
+// putValueToPeer stores the given key/value pair at the peer 'p'
+func (pm *ProtocolMessenger) PutValue(ctx context.Context, p peer.ID, rec *recpb.Record) error {
+	pmes := pb.NewMessage(pb.Message_PUT_VALUE, rec.Key, 0)
+	pmes.Record = rec
+	rpmes, err := pm.m.sendRequest(ctx, p, pmes)
+	if err != nil {
+		logger.Debugw("failed to put value to peer", "to", p, "key", loggableKeyBytes(rec.Key), "error", err)
+		return err
+	}
+
+	if !bytes.Equal(rpmes.GetRecord().Value, pmes.GetRecord().Value) {
+		logger.Infow("value not put correctly", "put-message", pmes, "get-message", rpmes)
+		return errors.New("value not put correctly")
+	}
+
+	return nil
+}
+
+// GetValue queries a particular peer p for the value for
+// key. It returns the value and a list of closer peers.
+func (pm *ProtocolMessenger) GetValue(ctx context.Context, p peer.ID, key string) (*recpb.Record, []*peer.AddrInfo, error) {
+	pmes := pb.NewMessage(pb.Message_GET_VALUE, []byte(key), 0)
+	respMsg, err := pm.m.sendRequest(ctx, p, pmes)
+	if err != nil {
+		return nil, nil, err
+	}
+
+	// Perhaps we were given closer peers
+	peers := pb.PBPeersToPeerInfos(respMsg.GetCloserPeers())
+
+	if record := respMsg.GetRecord(); record != nil {
+		// Success! We were given the value
+		logger.Debug("got value")
+
+		// make sure record is valid.
+		err = pm.validator.Validate(string(record.GetKey()), record.GetValue())
+		if err != nil {
+			logger.Debug("received invalid record (discarded)")
+			// return a sentinal to signify an invalid record was received
+			err = errInvalidRecord
+			record = new(recpb.Record)
+		}
+		return record, peers, err
+	}
+
+	if len(peers) > 0 {
+		return nil, peers, nil
+	}
+
+	return nil, nil, routing.ErrNotFound
+}
+
+// findPeerSingle asks peer 'p' if they know where the peer with id 'id' is
```

Comment:
GetClosestPeers asks p to return the K closest peers to id ?

Reply:
Similarly, it would be great to document all the exported functions here. Will be really helpful fot using this/understanding how the DHT API works.